soc: renesas: ra: ra4m1: Migrate to FSP-based configuration

soburi opened the pull request:
Change to use FSP to integrate with other Renesas RA series.

soburi committed:
soc: renesas: ra: ra4m1: Migrate to FSP-based configuration

Change to use FSP to integrate with other Renesas RA series.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

soburi committed:
drivers: clock_contrl: Remove renesas,ra-clock-generation-circuit driver

Remove the renesas,ra-clock-generation-circuit driver, which is no longer
needed after migrating to the FSP-based implementation.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

soburi committed:
dts: arm: renesas: ra4: Use renesas,ra-pinctrl-pfs driver

Switch the pinctrl driver to renesas,ra-pinctrl-pfs which can be
used with FSP.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

soburi committed:
drivers: pinctrl: Remove renesas,ra-pinctrl driver

Remove the renesas,ra-pinctrl driver, which is no longer
needed after migrating to the FSP-based implementation.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

soburi committed:
dts: arm: renesas: ra4: Use renesas,ra-cgc-pclkblock driver

Switch the clock controller driver to renesas,ra-cgc-pclkblock
which can be used with FSP.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

KhiemNguyenT reviewed (changes requested):
@soburi , I think you need to brush up the PR. There are some changes for RA2A1 are mixed into.

soburi committed:
soc: renesas: ra: ra4m1: Adapts the Option Setting Memory to FSP.

Since the Option Setting Memory area is set in FSP, the Kconfig value
switches between using the FSP implementation or the existing
Option Setting Memory implementation.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

soburi committed:
boards: mikroe: clicker_ra4m1: Migrate to FSP

Update configuration for migrate to FSP

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

soburi committed:
boards: arduino: uno_r4: Migrate to FSP

Update configuration for migrate to FSP

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

soburi commented:
> @soburi , I think you need to brush up the PR. There are some changes for RA2A1 are mixed into.

Update done.
I will remove DNM and proceed once the following PR in hal_renesas is approved.

https://github.com/zephyrproject-rtos/hal_renesas/pull/31

gmarull commented on dts/arm/renesas/ra/ra-cm4-common.dtsi:
```diff
@@ -21,49 +22,51 @@
 			#clock-cells = <0>;
+
+			/* PLL */
+			source = <RA_PLL_SOURCE_MAIN_OSC>;
+			div = <RA_PLL_DIV_2>;
```
wrong:

- source: define another clock and make it source using eg `clocks =  <&main_osc>;`
- divider should be given as integer `div = <2>;`.

soburi replied:
@duynguyenxa
Are these properties actually used?

duynguyenxa replied:
@soburi , these properties are used to provide information for the bsp_clock_cfg.h in the hal layer:
```
#define BSP_CFG_PLL_SOURCE                                                                         \
	BSP_CLOCK_PROP_HAS_STATUS_OKAY_OR(DT_NODELABEL(pll), source, RA_PLL_SOURCE_DISABLE)
#define BSP_CFG_PLL_DIV BSP_CLOCK_PROP_HAS_STATUS_OKAY_OR(DT_NODELABEL(pll), div, RA_PLL_DIV_2)
#if DT_NODE_HAS_STATUS(DT_NODELABEL(pll), okay)
```

The BSP_CFG_PLL_SOURCE, and BSP_CFG_PLL_DIV BSP will be used in the bsp_clock_init() to configure PLL clock.

gmarull replied:
FYI: We can't have HALisms in DT.

duynguyenxa replied:
@gmarull , the macro is defined in zephyr main repo, zephyr/include/zephyr/dt-bindings/clock/ra_clock.h

gmarull replied:
it still looks wrong, why isn't `div = <2>`? Are such values encoding what HAL expects?

duynguyenxa replied:
@gmarull , sorry for late response, it's expect the value map with HWM here for the pll.
![image](https://github.com/user-attachments/assets/ab410654-8351-4d46-8eba-a553568f7f3c)

soburi replied:
@duynguyenxa 

The root of this problem is that the DeviceTree is a hardware description language,
and how values ​​are handled should not be mixed in here.
In other words, the DeviceTree should be able to describe semantic values; for example, if the division factor is 2, then it should be describable as "2", and it should be up to the implementation side, i.e., the .h or .c file, to convert this to bit notation.

soburi replied:
I splitted out this issue to #78365.

soburi replied:
https://github.com/zephyrproject-rtos/zephyr/pull/78365 is already merged.

gmarull commented on dts/arm/renesas/ra/ra-cm4-common.dtsi:
```diff
@@ -82,19 +85,64 @@
+
+			pclka: pclka {
+				compatible = "renesas,ra-cgc-pclk";
+				clk_div = <RA_SYS_CLOCK_DIV_16>;
```
no underscores in properties please.

soburi replied:
I have created https://github.com/zephyrproject-rtos/zephyr/pull/78097.
I will fix it after this PR is merged.

soburi replied:
https://github.com/zephyrproject-rtos/zephyr/pull/78097 is already merged.

nordicjm commented on soc/renesas/ra/ra4m1/Kconfig:
```diff
@@ -5,10 +5,14 @@ config SOC_SERIES_RA4M1
 	select TIMER_READS_ITS_FREQUENCY_AT_RUNTIME
 	select XIP
+	select HAS_RENESAS_RA_FSP
+	help
+	  Enable support for Renesas RA4M1 MCU series
```
```suggestion
```

soburi replied:
Fixed it.

soburi committed:
boards: arduino: uno_r4: remove CONFIG_PINCTRL from defconfig of uno_r4

This PR fixes #78619 for the Arduino UNO R4 Minima/Wifi board.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

soburi commented:
@iandmorris I am looking to make part of the RA4M1 implementation FSP-based. Could you please review and check on your boards.

nordicjm reviewed (dismissed):
pinctrl and soc Kconfig/cmake changes OK

iandmorris commented:
> @iandmorris I am looking to make part of the RA4M1 implementation FSP-based. Could you please review and check on your boards.

@soburi - just confirmed that blinky still builds and runs on the Mikroe Clicker RA4M1

soburi commented:
> > @iandmorris I am looking to make part of the RA4M1 implementation FSP-based. Could you please review and check on your boards.
> 
> @soburi - just confirmed that blinky still builds and runs on the Mikroe Clicker RA4M1

Thank you for your confirmation. I would like to proceed with this.

soburi commented:
@KhiemNguyenT 
Could you take a look, please?

soburi commented:
@gmarull 
Could you please re-check your change request?

gmarull commented on soc/renesas/ra/ra4m1/soc.c:
```diff
@@ -131,6 +136,28 @@ const struct opt_set_mem ops __attribute__((section(".opt_set_mem"))) = {
+	return 0;
+}
+
+SYS_INIT(renesas_ra4m1_init, PRE_KERNEL_1, 0);
```
use hooks

soburi replied:
I fixed it.
Thank you for your review.

thaoluonguw reviewed (dismissed):
@soburi : It looks good to me for now.

soburi commented on boards/arduino/uno_r4/arduino_uno_r4_wifi-pinctrl.dtsi:
```diff
@@ -4,12 +4,18 @@
 	sci2_default: sci2_default {
 		group1 {
-			pinmux = <P301_RXD2>, <P302_TXD2>;
+			/* tx */
+			psels = <RA_PSEL(RA_PSEL_SCI_2, 3, 1)>;
```
Thank you. That's not good. I'll check it out.

soburi replied:
@thaoluonguw 

I rechecked it.
As you pointed out, this doesn't seem right, so I fixed it.

Here, RX  and TX were only swapped incorrectly.
So, both are the same configuration set to SCI2 function and do not behave incorrectly.
So, I failed to catch in the test.

soburi commented:
@KhiemNguyenT
Please approve if it is good for you.

tejlmand commented on soc/renesas/ra/ra4m1/CMakeLists.txt:
```diff
@@ -3,7 +3,7 @@
 zephyr_include_directories(.)
 
-zephyr_library_sources_ifdef(CONFIG_SOC_OPTION_SETTING_MEMORY
+zephyr_library_sources(
```
minor nit, not blocking.

No need for linewrap anymore.

soburi replied:
Thank you for your feedback.
I fixed it.

tejlmand reviewed (dismissed):
lgtm, a minot nit observed but not blocking.

soburi commented:
@KhiemNguyenT 
Please take a look.

KhiemNguyenT reviewed (approved):
Thank you for this action.
LGTM.
